syntax = "proto3";
package examples;
option go_package = "github.com/angzarr/angzarr/proto/examples;examples";
option java_package = "dev.angzarr.examples";
option java_multiple_files = true;
option csharp_namespace = "Angzarr.Examples";
option ruby_package = "Angzarr::Examples";

import "google/protobuf/timestamp.proto";

// =============================================================================
// Customer Domain - manages customer profiles and loyalty points
// =============================================================================

// Commands
message CreateCustomer {
  string name = 1;
  string email = 2;
}

message AddLoyaltyPoints {
  int32 points = 1;
  string reason = 2;  // e.g., "transaction:uuid" or "promotion:welcome"
}

message RedeemLoyaltyPoints {
  int32 points = 1;
  string redemption_type = 2;  // e.g., "discount", "reward"
}

// Events
message CustomerCreated {
  string name = 1;
  string email = 2;
  google.protobuf.Timestamp created_at = 3;
}

message LoyaltyPointsAdded {
  int32 points = 1;
  int32 new_balance = 2;
  string reason = 3;
}

message LoyaltyPointsRedeemed {
  int32 points = 1;
  int32 new_balance = 2;
  string redemption_type = 3;
}

// State (for snapshots)
message CustomerState {
  string name = 1;
  string email = 2;
  int32 loyalty_points = 3;
  int32 lifetime_points = 4;
}

// =============================================================================
// Product Domain - manages product catalog and pricing
// =============================================================================

// Commands
message CreateProduct {
  string sku = 1;
  string name = 2;
  string description = 3;
  int32 price_cents = 4;
}

message UpdateProduct {
  string name = 1;
  string description = 2;
}

message SetPrice {
  int32 price_cents = 1;
}

message Discontinue {
  string reason = 1;
}

// Events
message ProductCreated {
  string sku = 1;
  string name = 2;
  string description = 3;
  int32 price_cents = 4;
  google.protobuf.Timestamp created_at = 5;
}

message ProductUpdated {
  string name = 1;
  string description = 2;
  google.protobuf.Timestamp updated_at = 3;
}

message PriceSet {
  int32 price_cents = 1;
  int32 previous_price_cents = 2;
  google.protobuf.Timestamp set_at = 3;
}

message ProductDiscontinued {
  string reason = 1;
  google.protobuf.Timestamp discontinued_at = 2;
}

// State (for snapshots)
message ProductState {
  string sku = 1;
  string name = 2;
  string description = 3;
  int32 price_cents = 4;
  string status = 5;  // "active", "discontinued"
}

// =============================================================================
// Inventory Domain - manages stock levels and reservations
// =============================================================================

// Commands
message InitializeStock {
  string product_id = 1;
  int32 quantity = 2;
  int32 low_stock_threshold = 3;  // triggers LowStockAlert when available < threshold
}

message ReceiveStock {
  int32 quantity = 1;
  string reference = 2;  // e.g., purchase order number
}

message ReserveStock {
  int32 quantity = 1;
  string order_id = 2;
}

message ReleaseReservation {
  string order_id = 1;
}

message CommitReservation {
  string order_id = 1;
}

// Events
message StockInitialized {
  string product_id = 1;
  int32 quantity = 2;
  int32 low_stock_threshold = 3;
  google.protobuf.Timestamp initialized_at = 4;
}

message StockReceived {
  int32 quantity = 1;
  int32 new_on_hand = 2;
  string reference = 3;
  google.protobuf.Timestamp received_at = 4;
}

message StockReserved {
  int32 quantity = 1;
  string order_id = 2;
  int32 new_available = 3;
  google.protobuf.Timestamp reserved_at = 4;
}

message ReservationReleased {
  string order_id = 1;
  int32 quantity = 2;
  int32 new_available = 3;
  google.protobuf.Timestamp released_at = 4;
}

message ReservationCommitted {
  string order_id = 1;
  int32 quantity = 2;
  int32 new_on_hand = 3;
  google.protobuf.Timestamp committed_at = 4;
}

message LowStockAlert {
  string product_id = 1;
  int32 available = 2;
  int32 threshold = 3;
  google.protobuf.Timestamp alerted_at = 4;
}

// State (for snapshots)
message InventoryState {
  string product_id = 1;
  int32 on_hand = 2;
  int32 reserved = 3;
  int32 low_stock_threshold = 4;
  map<string, int32> reservations = 5;  // order_id -> quantity
}

// =============================================================================
// Order Domain - manages customer orders and payment
// =============================================================================

// Commands
message CreateOrder {
  string customer_id = 1;
  repeated LineItem items = 2;
}

message ApplyLoyaltyDiscount {
  int32 points = 1;           // loyalty points to use
  int32 discount_cents = 2;   // value of discount in cents
}

message SubmitPayment {
  string payment_method = 1;  // "card", "cash", etc.
  int32 amount_cents = 2;     // payment amount (must match total after discount)
}

message ConfirmPayment {
  string payment_reference = 1;  // external payment reference
}

message CancelOrder {
  string reason = 1;
}

// Events
message OrderCreated {
  string customer_id = 1;
  repeated LineItem items = 2;
  int32 subtotal_cents = 3;
  google.protobuf.Timestamp created_at = 4;
}

message LoyaltyDiscountApplied {
  int32 points_used = 1;
  int32 discount_cents = 2;
  google.protobuf.Timestamp applied_at = 3;
}

message PaymentSubmitted {
  string payment_method = 1;
  int32 amount_cents = 2;
  google.protobuf.Timestamp submitted_at = 3;
}

message OrderCompleted {
  int32 final_total_cents = 1;
  string payment_method = 2;
  string payment_reference = 3;
  int32 loyalty_points_earned = 4;  // Points to award customer (1 point per $1)
  google.protobuf.Timestamp completed_at = 5;
}

message OrderCancelled {
  string reason = 1;
  google.protobuf.Timestamp cancelled_at = 2;
  int32 loyalty_points_used = 3;
}

// Shared types
message LineItem {
  string product_id = 1;
  string name = 2;
  int32 quantity = 3;
  int32 unit_price_cents = 4;
}

// State (for snapshots)
message OrderState {
  string customer_id = 1;
  repeated LineItem items = 2;
  int32 subtotal_cents = 3;
  int32 discount_cents = 4;
  int32 loyalty_points_used = 5;
  string payment_method = 6;
  string payment_reference = 7;
  string status = 8;  // "pending", "payment_submitted", "completed", "cancelled"
}

// =============================================================================
// Cart Domain - manages shopping cart lifecycle
// =============================================================================

// Commands
message CreateCart {
  string customer_id = 1;
}

message AddItem {
  string product_id = 1;
  string name = 2;
  int32 quantity = 3;
  int32 unit_price_cents = 4;
}

message UpdateQuantity {
  string product_id = 1;
  int32 new_quantity = 2;
}

message RemoveItem {
  string product_id = 1;
}

message ApplyCoupon {
  string code = 1;
  string coupon_type = 2;  // "percentage", "fixed"
  int32 value = 3;         // percentage (0-100) or cents
}

message ClearCart {}

message Checkout {}

// Events
message CartCreated {
  string customer_id = 1;
  google.protobuf.Timestamp created_at = 2;
}

message ItemAdded {
  string product_id = 1;
  string name = 2;
  int32 quantity = 3;
  int32 unit_price_cents = 4;
  int32 new_subtotal = 5;
  google.protobuf.Timestamp added_at = 6;
}

message QuantityUpdated {
  string product_id = 1;
  int32 old_quantity = 2;
  int32 new_quantity = 3;
  int32 new_subtotal = 4;
  google.protobuf.Timestamp updated_at = 5;
}

message ItemRemoved {
  string product_id = 1;
  int32 quantity = 2;
  int32 new_subtotal = 3;
  google.protobuf.Timestamp removed_at = 4;
}

message CouponApplied {
  string coupon_code = 1;
  string coupon_type = 2;
  int32 value = 3;
  int32 discount_cents = 4;
  google.protobuf.Timestamp applied_at = 5;
}

message CartCleared {
  int32 new_subtotal = 1;
  google.protobuf.Timestamp cleared_at = 2;
}

message CartCheckedOut {
  int32 final_subtotal = 1;
  int32 discount_cents = 2;
  google.protobuf.Timestamp checked_out_at = 3;
}

// State (for snapshots)
message CartItem {
  string product_id = 1;
  string name = 2;
  int32 quantity = 3;
  int32 unit_price_cents = 4;
}

message CartState {
  string customer_id = 1;
  repeated CartItem items = 2;
  int32 subtotal_cents = 3;
  string coupon_code = 4;
  int32 discount_cents = 5;
  string status = 6;  // "active", "checked_out"
}

// =============================================================================
// Fulfillment Domain - manages shipment lifecycle
// =============================================================================

// Commands
message CreateShipment {
  string order_id = 1;
}

message MarkPicked {
  string picker_id = 1;
}

message MarkPacked {
  string packer_id = 1;
}

message Ship {
  string carrier = 1;
  string tracking_number = 2;
}

message RecordDelivery {
  string signature = 1;
}

// Events
message ShipmentCreated {
  string order_id = 1;
  string status = 2;
  google.protobuf.Timestamp created_at = 3;
}

message ItemsPicked {
  string picker_id = 1;
  google.protobuf.Timestamp picked_at = 2;
}

message ItemsPacked {
  string packer_id = 1;
  google.protobuf.Timestamp packed_at = 2;
}

message Shipped {
  string carrier = 1;
  string tracking_number = 2;
  google.protobuf.Timestamp shipped_at = 3;
}

message Delivered {
  string signature = 1;
  google.protobuf.Timestamp delivered_at = 2;
}

// State (for snapshots)
message FulfillmentState {
  string order_id = 1;
  string status = 2;  // "pending", "picking", "packing", "shipped", "delivered"
  string tracking_number = 3;
  string carrier = 4;
  string picker_id = 5;
  string packer_id = 6;
  string signature = 7;
}

// =============================================================================
// Projections
// =============================================================================

// Receipt projection generated when an order completes
message Receipt {
  string order_id = 1;
  string customer_id = 2;
  repeated LineItem items = 3;
  int32 subtotal_cents = 4;
  int32 discount_cents = 5;
  int32 final_total_cents = 6;
  string payment_method = 7;
  int32 loyalty_points_earned = 8;
  google.protobuf.Timestamp completed_at = 9;
  string formatted_text = 10;  // Human-readable receipt text
}
